### PR TITLE
Added Harvest Source for new time entries

### PR DESCRIPTION
**Pull request by justabuzz**
Added a new Source for Harvest app, watching for new time entries.

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4130"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import harvest from "../../harvest.app.mjs";
+
+export default {
+  key: "harvest-new-timeentry-entry",
```
Comment (lcaresia) [inline components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:4]: The correct isn't `timeentry-entry`?

If yes, update the name of this file too!

```suggestion
  key: "harvest-new-time-entry",
```

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  version: "0.0.1",
+  type: "source",
+  props: {
+    harvest,
+    timer: {
+      label: "Polling interval",
+      description: "Pipedream will poll Harvest API on this schedule",
+      type: "$.interface.timer",
+      default: {
+        intervalSeconds: 60 * 15,
```
Comment (lcaresia) [inline components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:16]:
```suggestion
        intervalSeconds: 60 * 15, // 15 minutes
```

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    let lastDateChecked = this.harvest.getLastDateChecked(this.db);
+    if (!lastDateChecked) {
+      lastDateChecked = new Date().toISOString();
+      this.harvest.setLastDateChecked(this.db, lastDateChecked);
+    }
+    const time_entries = await this.harvest.listTimeEntriesPaginated({
+      page: 1,
+      updatedSince: lastDateChecked,
+    });
+    for await (const time_entry of time_entries) {
```
Comment (lcaresia) [inline components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:33]: I think don't need this `await` here.

Obs: I din't tested.

```suggestion
    for (const time_entry of time_entries) {
```

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      data.push(time_entry);
+    }
+    data && data.reverse().forEach((time_entry) => {
+      this.harvest.setLastDateChecked(this.db, time_entry.updated_at);
+      this.$emit(time_entry,
+        {
+          id: time_entry.id,
+          summary: `Time Entry id: ${time_entry.id}`,
```
Comment (lcaresia) [inline components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:41]: Is missing ts prop.

```suggestion
          summary: `Time Entry id: ${time_entry.id}`,
          ts: Date.parse(time_entry.updated_at), // or new Date()
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Just some fixes!

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs (hunk 1 above)
Comment (justabuzz) [reply components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:4]: Yeah, it sounds funny, but it's a [Time Entry](https://help.getharvest.com/api-v2/timesheets-api/timesheets/time-entries/) object. So we are watching for new Time Entry entries :)

File: components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs (hunk 3 above)
Comment (justabuzz) [reply components/harvest/sources/new-timeentry-entry/new-timeentry-entry.mjs:33]: Hi, await is required. Tested.

**Comment by justabuzz** [comment]
Hi, I have addressed the feedback and submitted a request for another review from @lcaresia. Do I need to do anything else now? Or just wait? Thanks!

**Comment by dannyroosevelt** [comment]
@justabuzz sorry for the delay! @lcaresia is going to take another look asap. Have you been able to successfully test the source? Happy to publish this asap if so, otherwise our QA team can check it out beforehand.

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by justabuzz** [comment]
> 

Thanks @dannyroosevelt, yes it's tested and works like a charm under My Actions for the past few days. I saw that @lcaresia already approved it, so we are good. Thanks again everyone and glad I can give back!

**Comment by dannyroosevelt** [comment]
> Thanks @dannyroosevelt, yes it's tested and works like a charm under my account for the past few days. I saw that @lcaresia already approved it, so we are good. Thanks again everyone and glad I can give back!

Great! I just merged your PR, so the new source is available to all Pipedream users now! Thanks a ton for your contribution! https://pipedream.com/apps/harvest#popular-harvest-triggers

**Pull request merged**